pshtt.py

```diff
@@ -402,29 +418,30 @@ def canonical_endpoint(http, httpwww, https, httpswww):
     # It allows a site to be canonically HTTPS if the cert has
     # a valid hostname but invalid chain issues.
 
-    is_https = (
-        (
-            (https.live and (not https.https_bad_hostname)) or
-            (httpswww.live and (not https.https_bad_hostname))
-        ) and (
-            (
-                http.redirect or
-                (not http.live) or
-                (not str(http.status).startswith("2"))
-            ) and (
-                httpwww.redirect or
-                (not httpwww.live) or
-                (not str(httpwww.status).startswith("2"))
-            )
-        ) and (
-            (
-                http.redirect_immediately_to_https and
-                (not http.redirect_immediately_to_external)
-            ) or (
-                httpwww.redirect_immediately_to_https and
-                (not httpwww.redirect_immediately_to_external)
-            )
+    def https_used(endpoint):
+        return endpoint.live and (not endpoint.https_bad_hostname)
+
+    def http_unused(endpoint):
+        return (
+            endpoint.redirect or
+            (not endpoint.live) or
+            (not str(endpoint.status).startswith("2"))
         )
+
+    def http_upgrades(endpoint):
+        return (
+            endpoint.redirect_immediately_to_https and
+            (not endpoint.redirect_immediately_to_external)
+        )
+
+    at_least_one_https_endpoint = https_used(https) or https_used(httpswww)
+    all_http_unused = http_unused(http) and http_unused(httpwww)
+    at_least_one_http_upgrades = http_upgrades(http) or http_upgrades(httpwww)
+
+    is_https = (
+        at_least_one_https_endpoint and
+        all_http_unused and
+        at_least_one_http_upgrades
```

Comment:
This refactored approach to calculating whether a domain's canonical endpoint uses HTTPS should be identical to the intended logic before, but should be easier to read, understand, and debug.